File: hu.bme.mit.theta.core/src/main/java/hu/bme/mit/theta/core/type/inttype/IntNeqExpr.java
package hu.bme.mit.theta.core.type.inttype;

import static hu.bme.mit.theta.core.type.booltype.BoolExprs.Bool;

import hu.bme.mit.theta.core.Expr;
import hu.bme.mit.theta.core.model.Substitution;
import hu.bme.mit.theta.core.type.abstracttype.NeqExpr;
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr;
import hu.bme.mit.theta.core.type.booltype.BoolType;

public final class IntNeqExpr extends NeqExpr<IntType> {

	private static final int HASH_SEED = 9397;
	private static final String OPERATOR_LABEL = "Neq";

	IntNeqExpr(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		super(leftOp, rightOp);
	}

	@Override
	public BoolType getType() {
		return Bool();
	}

	@Override
	public BoolLitExpr eval(final Substitution assignment) {
		final IntLitExpr leftOpVal = (IntLitExpr) getLeftOp().eval(assignment);
		final IntLitExpr rightOpVal = (IntLitExpr) getRightOp().eval(assignment);
		return leftOpVal.neq(rightOpVal);
	}

	@Override
	public IntNeqExpr with(final Expr<IntType> leftOp, final Expr<IntType> rightOp) {
		if (leftOp == getLeftOp() && rightOp == getRightOp()) {
			return this;
		} else {
			return new IntNeqExpr(leftOp, rightOp);
		}
	}

	@Override
	public IntNeqExpr withLeftOp(final Expr<IntType> leftOp) {
		return with(leftOp, getRightOp());
	}

	@Override
	public IntNeqExpr withRightOp(final Expr<IntType> rightOp) {
		return with(getLeftOp(), rightOp);
	}

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		} else if (obj instanceof IntNeqExpr) {
			final IntNeqExpr that = (IntNeqExpr) obj;
			return this.getLeftOp().equals(that.getLeftOp()) && this.getRightOp().equals(that.getRightOp());
		} else {
			return false;
		}
	}

	@Override
	protected int getHashSeed() {
		return HASH_SEED;
	}

	@Override
	protected String getOperatorLabel() {
		return OPERATOR_LABEL;
	}

}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/inttype/IntNeqExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hu.bme.mit.theta.core.type.inttype

import hu.bme.mit.theta.core.model.Valuation
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.type.abstracttype.NeqExpr
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr
import hu.bme.mit.theta.core.type.inttype.IntExprs.Int
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@SerialName("IntNeq")
data class IntNeqExpr(
    override val leftOp: Expr<IntType>,
    override val rightOp: Expr<IntType>
) : NeqExpr<IntType>() {

    companion object {

        @JvmStatic
        fun of(leftOp: Expr<IntType>, rightOp: Expr<IntType>) = IntNeqExpr(leftOp, rightOp)
        @JvmStatic
        fun create(leftOp: Expr<*>, rightOp: Expr<*>) = IntNeqExpr(cast(leftOp, Int()), cast(rightOp, Int()))
    }

    override fun eval(`val`: Valuation): BoolLitExpr {
        val leftOpVal = leftOp.eval(`val`) as IntLitExpr
        val rightOpVal = rightOp.eval(`val`) as IntLitExpr
        return leftOpVal.neq(rightOpVal)
    }

    override fun of(leftOp: Expr<IntType>, rightOp: Expr<IntType>): IntNeqExpr =
        Companion.of(leftOp, rightOp)

    override fun toString(): String = super.toString()
}

